File: submissions/shashiguru-keluth/java/src/test/java/com/banking/BankingClientTest.java
package com.banking;

import com.banking.config.BankingClientConfig;
import com.banking.model.TransferRequest;
import com.banking.model.TransferResponse;
import com.google.gson.Gson;
import com.google.gson.JsonObject;
import org.junit.jupiter.api.*;
import org.mockito.MockedStatic;

import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.List;
import java.util.Optional;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.*;

/**
 * Unit tests for BankingClient
 * Tests all functionality with mocked HTTP responses
 */
class BankingClientTest {

    private BankingClient client;
    private BankingClientConfig config;
    private HttpClient mockHttpClient;
    private static final Gson gson = new Gson();

    @BeforeEach
    void setUp() {
        config = new BankingClientConfig.Builder()
                .baseUrl("http://localhost:8123")
                .timeout(30)
                .maxRetries(3)
                .build();
        client = new BankingClient(config);
    }

    @AfterEach
    void tearDown() {
        if (client != null) {
            client.close();
        }
    }

    @Nested
    @DisplayName("Constructor Tests")
    class ConstructorTests {
        
        @Test
        @DisplayName("Should initialize with default config")
        void testDefaultConstructor() {
            BankingClient defaultClient = new BankingClient();
            assertNotNull(defaultClient);
            defaultClient.close();
        }

        @Test
        @DisplayName("Should initialize with custom config")
        void testCustomConstructor() {
            BankingClientConfig customConfig = new BankingClientConfig.Builder()
                    .baseUrl("http://example.com")
                    .build();
            BankingClient customClient = new BankingClient(customConfig);
            assertNotNull(customClient);
            customClient.close();
        }
    }

    @Nested
    @DisplayName("Authentication Tests")
    class AuthenticationTests {

        @Test
        @DisplayName("Should authenticate successfully")
        void testAuthenticateSuccess() {
            // Note: This is a simplified test. In production, you'd mock HttpClient
            // For demonstration, we're testing the structure
            assertDoesNotThrow(() -> client.authenticate("testuser", "password"));
        }

        @Test
        @DisplayName("Should handle authentication with empty credentials")
        void testAuthenticateEmptyCredentials() {
            assertDoesNotThrow(() -> client.authenticate("", ""));
        }
    }

    @Nested
    @DisplayName("Account Validation Tests")
    class AccountValidationTests {

        @Test
        @DisplayName("Should validate account")
        void testValidateAccount() {
            assertDoesNotThrow(() -> client.validateAccount("ACC1000"));
        }

        @Test
        @DisplayName("Should handle null account ID")
        void testValidateNullAccount() {
            assertDoesNotThrow(() -> client.validateAccount(null));
        }
    }

    @Nested
    @DisplayName("Transfer Tests")
    class TransferTests {

        @Test
        @DisplayName("Should transfer funds without authentication")
        void testTransferFundsWithoutAuth() {
            Optional<TransferResponse> result = client.transferFunds(
                    "ACC1000", "ACC1001", 100.0, false
            );
            // Result depends on server availability
            assertNotNull(result);
        }

        @Test
        @DisplayName("Should transfer funds with authentication")
        void testTransferFundsWithAuth() {
            Optional<TransferResponse> result = client.transferFunds(
                    "ACC1000", "ACC1001", 100.0, true
            );
            assertNotNull(result);
        }

        @Test
        @DisplayName("Should handle invalid amount")
        void testTransferInvalidAmount() {
            Optional<TransferResponse> result = client.transferFunds(
                    "ACC1000", "ACC1001", -100.0, false
            );
            assertNotNull(result);
            assertTrue(result.isEmpty());
        }

        @Test
        @DisplayName("Should handle zero amount")
        void testTransferZeroAmount() {
            Optional<TransferResponse> result = client.transferFunds(
                    "ACC1000", "ACC1001", 0.0, false
            );
            assertNotNull(result);
            assertTrue(result.isEmpty());
        }

        @Test
        @DisplayName("Should handle empty from account")
        void testTransferEmptyFromAccount() {
            Optional<TransferResponse> result = client.transferFunds(
                    "", "ACC1001", 100.0, false
            );
            assertNotNull(result);
            assertTrue(result.isEmpty());
        }

        @Test
        @DisplayName("Should handle empty to account")
        void testTransferEmptyToAccount() {
            Optional<TransferResponse> result = client.transferFunds(
                    "ACC1000", "", 100.0, false
            );
            assertNotNull(result);
            assertTrue(result.isEmpty());
        }
    }

    @Nested
    @DisplayName("Get Accounts Tests")
    class GetAccountsTests {

        @Test
        @DisplayName("Should get accounts without authentication")
        void testGetAccountsWithoutAuth() {
            Optional<List<JsonObject>> result = client.getAccounts(false);
            assertNotNull(result);
        }

        @Test
        @DisplayName("Should get accounts with authentication")
        void testGetAccountsWithAuth() {
            Optional<List<JsonObject>> result = client.getAccounts(true);
            assertNotNull(result);
        }
    }

    @Nested
    @DisplayName("Get Account Balance Tests")
    class GetAccountBalanceTests {

        @Test
        @DisplayName("Should get account balance")
        void testGetAccountBalance() {
            Optional<JsonObject> result = client.getAccountBalance("ACC1000", false);
            assertNotNull(result);
        }

        @Test
        @DisplayName("Should handle null account ID")
        void testGetBalanceNullAccount() {
            assertDoesNotThrow(() -> client.getAccountBalance(null, false));
        }
    }

    @Nested
    @DisplayName("TransferRequest Tests")
    class TransferRequestTests {

        @Test
        @DisplayName("Should create valid transfer request")
        void testValidTransferRequest() {
            TransferRequest request = new TransferRequest.Builder()
                    .fromAccount("ACC1000")
                    .toAccount("ACC1001")
                    .amount(100.0)
                    .build();

            assertEquals("ACC1000", request.getFromAccount());
            assertEquals("ACC1001", request.getToAccount());
            assertEquals(100.0, request.getAmount());
        }

        @Test
        @DisplayName("Should throw exception for null from account")
        void testNullFromAccount() {
            assertThrows(IllegalArgumentException.class, () ->
                    new TransferRequest.Builder()
                            .fromAccount(null)
                            .toAccount("ACC1001")
                            .amount(100.0)
                            .build()
            );
        }

        @Test
        @DisplayName("Should throw exception for empty from account")
        void testEmptyFromAccount() {
            assertThrows(IllegalArgumentException.class, () ->
                    new TransferRequest.Builder()
                            .fromAccount("")
                            .toAccount("ACC1001")
                            .amount(100.0)
                            .build()
            );
        }

        @Test
        @DisplayName("Should throw exception for null to account")
        void testNullToAccount() {
            assertThrows(IllegalArgumentException.class, () ->
                    new TransferRequest.Builder()
                            .fromAccount("ACC1000")
                            .toAccount(null)
                            .amount(100.0)
                            .build()
            );
        }

        @Test
        @DisplayName("Should throw exception for zero amount")
        void testZeroAmount() {
            assertThrows(IllegalArgumentException.class, () ->
                    new TransferRequest.Builder()
                            .fromAccount("ACC1000")
                            .toAccount("ACC1001")
                            .amount(0.0)
                            .build()
            );
        }

        @Test
        @DisplayName("Should throw exception for negative amount")
        void testNegativeAmount() {
            assertThrows(IllegalArgumentException.class, () ->
                    new TransferRequest.Builder()
                            .fromAccount("ACC1000")
                            .toAccount("ACC1001")
                            .amount(-100.0)
                            .build()
            );
        }

        @Test
        @DisplayName("Should throw exception when from account not set")
        void testFromAccountNotSet() {
            assertThrows(IllegalStateException.class, () ->
                    new TransferRequest.Builder()
                            .toAccount("ACC1001")
                            .amount(100.0)
                            .build()
            );
        }

        @Test
        @DisplayName("Should throw exception when to account not set")
        void testToAccountNotSet() {
            assertThrows(IllegalStateException.class, () ->
                    new TransferRequest.Builder()
                            .fromAccount("ACC1000")
                            .amount(100.0)
                            .build()
            );
        }
    }

    @Nested
    @DisplayName("TransferResponse Tests")
    class TransferResponseTests {

        @Test
        @DisplayName("Should create transfer response")
        void testCreateTransferResponse() {
            TransferResponse response = new TransferResponse(
                    "txn_123",
                    "SUCCESS",
                    "Transfer completed",
                    "ACC1000",
                    "ACC1001",
                    100.0
            );

            assertEquals("txn_123", response.getTransactionId());
            assertEquals("SUCCESS", response.getStatus());
            assertEquals("Transfer completed", response.getMessage());
            assertEquals("ACC1000", response.getFromAccount());
            assertEquals("ACC1001", response.getToAccount());
            assertEquals(100.0, response.getAmount());
        }

        @Test
        @DisplayName("Should have toString method")
        void testToString() {
            TransferResponse response = new TransferResponse(
                    "txn_123",
                    "SUCCESS",
                    "Transfer completed",
                    "ACC1000",
                    "ACC1001",
                    100.0
            );

            String str = response.toString();
            assertNotNull(str);
            assertTrue(str.contains("txn_123"));
            assertTrue(str.contains("SUCCESS"));
        }
    }

    @Nested
    @DisplayName("BankingClientConfig Tests")
    class BankingClientConfigTests {

        @Test
        @DisplayName("Should build with default values")
        void testDefaultConfig() {
            BankingClientConfig config = new BankingClientConfig.Builder().build();

            assertEquals("http://localhost:8123", config.getBaseUrl());
            assertEquals(30, config.getTimeout());
            assertEquals(3, config.getMaxRetries());
        }

        @Test
        @DisplayName("Should build with custom values")
        void testCustomConfig() {
            BankingClientConfig config = new BankingClientConfig.Builder()
                    .baseUrl("http://example.com:9000")
                    .timeout(60)
                    .maxRetries(5)
                    .retryDelay(2000)
                    .build();

            assertEquals("http://example.com:9000", config.getBaseUrl());
            assertEquals(60, config.getTimeout());
            assertEquals(5, config.getMaxRetries());
            assertEquals(2000, config.getRetryDelay());
        }

        @Test
        @DisplayName("Should remove trailing slash from base URL")
        void testTrailingSlashRemoval() {
            BankingClientConfig config = new BankingClientConfig.Builder()
                    .baseUrl("http://example.com/")
                    .build();

            assertEquals("http://example.com", config.getBaseUrl());
        }

        @Test
        @DisplayName("Should throw exception for invalid timeout")
        void testInvalidTimeout() {
            assertThrows(IllegalArgumentException.class, () ->
                    new BankingClientConfig.Builder()
                            .timeout(-1)
                            .build()
            );
        }

        @Test
        @DisplayName("Should throw exception for negative max retries")
        void testNegativeMaxRetries() {
            assertThrows(IllegalArgumentException.class, () ->
                    new BankingClientConfig.Builder()
                            .maxRetries(-1)
                            .build()
            );
        }

        @Test
        @DisplayName("Should throw exception for invalid retry delay")
        void testInvalidRetryDelay() {
            assertThrows(IllegalArgumentException.class, () ->
                    new BankingClientConfig.Builder()
                            .retryDelay(-1)
                            .build()
            );
        }

        @Test
        @DisplayName("Should allow zero max retries")
        void testZeroMaxRetries() {
            assertDoesNotThrow(() ->
                    new BankingClientConfig.Builder()
                            .maxRetries(0)
                            .build()
            );
        }
    }

    @Nested
    @DisplayName("Integration Tests")
    class IntegrationTests {

        @Test
        @DisplayName("Should handle full authentication and transfer flow")
        void testFullFlow() {
            // This test demonstrates the expected flow
            // In production, this would use a mock server
            assertDoesNotThrow(() -> {
                client.authenticate("testuser", "password");
                client.validateAccount("ACC1000");
                client.transferFunds("ACC1000", "ACC1001", 100.0, false);
            });
        }

        @Test
        @DisplayName("Should handle multiple operations")
        void testMultipleOperations() {
            assertDoesNotThrow(() -> {
                client.getAccounts(false);
                client.validateAccount("ACC1000");
                client.getAccountBalance("ACC1000", false);
            });
        }
    }

    @Nested
    @DisplayName("Resource Management Tests")
    class ResourceManagementTests {

        @Test
        @DisplayName("Should close properly")
        void testClose() {
            BankingClient testClient = new BankingClient();
            assertDoesNotThrow(testClient::close);
        }

        @Test
        @DisplayName("Should work with try-with-resources")
        void testTryWithResources() {
            assertDoesNotThrow(() -> {
                try (BankingClient autoClient = new BankingClient()) {
                    autoClient.validateAccount("ACC1000");
                }
            });
        }
    }
}

File: submissions/shashiguru-keluth/javascript/bankingClient.test.js
/**
 * Unit tests for Modern Banking Client (JavaScript)
 * Tests all functionality with mocked fetch responses
 */

// Mock fetch globally
global.fetch = jest.fn();

const { BankingClient, BankingClientConfig, TransferRequest, Logger } = require('./bankingClient');

describe('TransferRequest', () => {
  describe('Valid requests', () => {
    test('should create valid transfer request', () => {
      const request = new TransferRequest('ACC1000', 'ACC1001', 100.00);
      expect(request.fromAccount).toBe('ACC1000');
      expect(request.toAccount).toBe('ACC1001');
      expect(request.amount).toBe(100.00);
    });

    test('should convert to JSON correctly', () => {
      const request = new TransferRequest('ACC1000', 'ACC1001', 100.00);
      const json = request.toJSON();
      expect(json).toEqual({
        fromAccount: 'ACC1000',
        toAccount: 'ACC1001',
        amount: 100.00
      });
    });
  });

  describe('Invalid requests', () => {
    test('should throw error for empty fromAccount', () => {
      expect(() => new TransferRequest('', 'ACC1001', 100))
        .toThrow('Invalid fromAccount');
    });

    test('should throw error for empty toAccount', () => {
      expect(() => new TransferRequest('ACC1000', '', 100))
        .toThrow('Invalid toAccount');
    });

    test('should throw error for zero amount', () => {
      expect(() => new TransferRequest('ACC1000', 'ACC1001', 0))
        .toThrow('Invalid amount');
    });

    test('should throw error for negative amount', () => {
      expect(() => new TransferRequest('ACC1000', 'ACC1001', -100))
        .toThrow('Invalid amount');
    });

    test('should throw error for non-string fromAccount', () => {
      expect(() => new TransferRequest(null, 'ACC1001', 100))
        .toThrow('Invalid fromAccount');
    });
  });
});

describe('BankingClientConfig', () => {
  test('should have default values', () => {
    const config = new BankingClientConfig();
    expect(config.baseUrl).toBe('http://localhost:8123');
    expect(config.timeout).toBe(30000);
    expect(config.maxRetries).toBe(3);
  });

  test('should accept custom values', () => {
    const config = new BankingClientConfig({
      baseUrl: 'http://example.com:9000',
      timeout: 60000,
      maxRetries: 5,
      retryDelay: 2000
    });
    expect(config.baseUrl).toBe('http://example.com:9000');
    expect(config.timeout).toBe(60000);
    expect(config.maxRetries).toBe(5);
    expect(config.retryDelay).toBe(2000);
  });

  test('should remove trailing slash from baseUrl', () => {
    const config = new BankingClientConfig({ baseUrl: 'http://example.com/' });
    expect(config.baseUrl).toBe('http://example.com');
  });
});

describe('Logger', () => {
  let consoleLogSpy;

  beforeEach(() => {
    consoleLogSpy = jest.spyOn(console, 'log').mockImplementation();
  });

  afterEach(() => {
    consoleLogSpy.mockRestore();
  });

  test('should log info messages', () => {
    Logger.info('Test message');
    expect(consoleLogSpy).toHaveBeenCalled();
  });

  test('should log error messages', () => {
    Logger.error('Error message');
    expect(consoleLogSpy).toHaveBeenCalled();
  });

  test('should log with data', () => {
    Logger.info('Test', { key: 'value' });
    expect(consoleLogSpy).toHaveBeenCalledTimes(2); // Once for message, once for data
  });
});

describe('BankingClient', () => {
  let client;
  
  beforeEach(() => {
    jest.clearAllMocks();
    client = new BankingClient();
  });

  describe('Constructor', () => {
    test('should initialize with default config', () => {
      expect(client.config).toBeDefined();
      expect(client.jwtToken).toBeNull();
    });

    test('should initialize with custom config', () => {
      const config = new BankingClientConfig({ baseUrl: 'http://custom.com' });
      const customClient = new BankingClient(config);
      expect(customClient.config.baseUrl).toBe('http://custom.com');
    });
  });

  describe('authenticate', () => {
    test('should authenticate successfully', async () => {
      const mockResponse = {
        ok: true,
        status: 200,
        json: async () => ({ token: 'test_token_123' })
      };
      fetch.mockResolvedValueOnce(mockResponse);

      const result = await client.authenticate('testuser', 'password');

      expect(result).toBe(true);
      expect(client.jwtToken).toBe('test_token_123');
      expect(fetch).toHaveBeenCalledWith(
        expect.stringContaining('/authToken'),
        expect.objectContaining({ method: 'POST' })
      );
    });

    test('should handle authentication failure', async () => {
      const mockResponse = {
        ok: false,
        status: 401,
        statusText: 'Unauthorized'
      };
      fetch.mockResolvedValueOnce(mockResponse);

      const result = await client.authenticate('testuser', 'wrongpassword');

      expect(result).toBe(false);
      expect(client.jwtToken).toBeNull();
    });

    test('should handle network error', async () => {
      fetch.mockRejectedValueOnce(new Error('Network error'));

      const result = await client.authenticate('testuser', 'password');

      expect(result).toBe(false);
    });

    test('should handle missing token in response', async () => {
      const mockResponse = {
        ok: true,
        status: 200,
        json: async () => ({})
      };
      fetch.mockResolvedValueOnce(mockResponse);

      const result = await client.authenticate('testuser', 'password');

      expect(result).toBe(false);
    });
  });

  describe('validateAccount', () => {
    test('should validate valid account', async () => {
      const mockResponse = {
        ok: true,
        status: 200,
        json: async () => ({ valid: true })
      };
      fetch.mockResolvedValueOnce(mockResponse);

      const result = await client.validateAccount('ACC1000');

      expect(result).toBe(true);
    });

    test('should invalidate invalid account', async () => {
      const mockResponse = {
        ok: true,
        status: 200,
        json: async () => ({ valid: false })
      };
      fetch.mockResolvedValueOnce(mockResponse);

      const result = await client.validateAccount('ACC9999');

      expect(result).toBe(false);
    });

    test('should handle validation error', async () => {
      fetch.mockRejectedValueOnce(new Error('Network error'));

      const result = await client.validateAccount('ACC1000');

      expect(result).toBe(false);
    });
  });

  describe('transferFunds', () => {
    test('should transfer funds successfully', async () => {
      const mockResponse = {
        ok: true,
        status: 200,
        json: async () => ({
          transactionId: 'txn_123',
          status: 'SUCCESS',
          message: 'Transfer completed',
          fromAccount: 'ACC1000',
          toAccount: 'ACC1001',
          amount: 100.00
        })
      };
      fetch.mockResolvedValueOnce(mockResponse);

      const result = await client.transferFunds('ACC1000', 'ACC1001', 100.00);

      expect(result).not.toBeNull();
      expect(result.transactionId).toBe('txn_123');
      expect(result.status).toBe('SUCCESS');
      expect(result.amount).toBe(100.00);
    });

    test('should transfer with authentication', async () => {
      client.jwtToken = 'test_token';
      const mockResponse = {
        ok: true,
        status: 200,
        json: async () => ({
          transactionId: 'txn_456',
          status: 'SUCCESS',
          message: 'Transfer completed',
          fromAccount: 'ACC1000',
          toAccount: 'ACC1001',
          amount: 100.00
        })
      };
      fetch.mockResolvedValueOnce(mockResponse);

      const result = await client.transferFunds('ACC1000', 'ACC1001', 100.00, true);

      expect(result).not.toBeNull();
      expect(fetch).toHaveBeenCalledWith(
        expect.any(String),
        expect.objectContaining({
          headers: expect.objectContaining({
            'Authorization': 'Bearer test_token'
          })
        })
      );
    });

    test('should handle invalid transfer data', async () => {
      const result = await client.transferFunds('', 'ACC1001', 100.00);

      expect(result).toBeNull();
      expect(fetch).not.toHaveBeenCalled();
    });

    test('should handle transfer error', async () => {
      const mockResponse = {
        ok: false,
        status: 400,
        statusText: 'Bad Request',
        json: async () => ({ message: 'Invalid account' })
      };
      fetch.mockResolvedValueOnce(mockResponse);

      const result = await client.transferFunds('ACC9999', 'ACC1001', 100.00);

      expect(result).toBeNull();
    });

    test('should handle network error during transfer', async () => {
      fetch.mockRejectedValueOnce(new Error('Network error'));

      const result = await client.transferFunds('ACC1000', 'ACC1001', 100.00);

      expect(result).toBeNull();
    });
  });

  describe('getAccounts', () => {
    test('should get accounts successfully', async () => {
      const mockAccounts = [
        { accountId: 'ACC1000', accountHolder: 'John Doe' },
        { accountId: 'ACC1001', accountHolder: 'Jane Smith' }
      ];
      const mockResponse = {
        ok: true,
        status: 200,
        json: async () => mockAccounts
      };
      fetch.mockResolvedValueOnce(mockResponse);

      const result = await client.getAccounts();

      expect(result).not.toBeNull();
      expect(result.length).toBe(2);
      expect(result[0].accountId).toBe('ACC1000');
    });

    test('should get accounts with authentication', async () => {
      client.jwtToken = 'test_token';
      const mockResponse = {
        ok: true,
        status: 200,
        json: async () => []
      };
      fetch.mockResolvedValueOnce(mockResponse);

      const result = await client.getAccounts(true);

      expect(result).not.toBeNull();
      expect(fetch).toHaveBeenCalledWith(
        expect.any(String),
        expect.objectContaining({
          headers: expect.objectContaining({
            'Authorization': 'Bearer test_token'
          })
        })
      );
    });

    test('should handle error getting accounts', async () => {
      fetch.mockRejectedValueOnce(new Error('Network error'));

      const result = await client.getAccounts();

      expect(result).toBeNull();
    });
  });

  describe('getAccountBalance', () => {
    test('should get balance successfully', async () => {
      const mockResponse = {
        ok: true,
        status: 200,
        json: async () => ({
          accountId: 'ACC1000',
          balance: 5000.00,
          currency: 'USD'
        })
      };
      fetch.mockResolvedValueOnce(mockResponse);

      const result = await client.getAccountBalance('ACC1000');

      expect(result).not.toBeNull();
      expect(result.accountId).toBe('ACC1000');
      expect(result.balance).toBe(5000.00);
    });

    test('should handle error getting balance', async () => {
      fetch.mockRejectedValueOnce(new Error('Network error'));

      const result = await client.getAccountBalance('ACC1000');

      expect(result).toBeNull();
    });
  });

  describe('_getHeaders', () => {
    test('should return headers without auth', () => {
      const headers = client._getHeaders(false);

      expect(headers['Content-Type']).toBe('application/json');
      expect(headers['Authorization']).toBeUndefined();
    });

    test('should return headers with auth', () => {
      client.jwtToken = 'test_token_123';
      const headers = client._getHeaders(true);

      expect(headers['Content-Type']).toBe('application/json');
      expect(headers['Authorization']).toBe('Bearer test_token_123');
    });

    test('should not include auth if no token available', () => {
      const headers = client._getHeaders(true);

      expect(headers['Authorization']).toBeUndefined();
    });
  });

  describe('_fetchWithTimeout', () => {
    test('should call fetch with correct parameters', async () => {
      const mockResponse = { ok: true, status: 200 };
      fetch.mockResolvedValueOnce(mockResponse);

      const result = await client._fetchWithTimeout('http://example.com', { method: 'GET' });

      expect(result).toEqual(mockResponse);
      expect(fetch).toHaveBeenCalledWith(
        'http://example.com',
        expect.objectContaining({ method: 'GET' })
      );
    });
  });

  describe('_fetchWithRetry', () => {
    test('should succeed on first try', async () => {
      fetch.mockResolvedValueOnce({ ok: true, status: 200, json: async () => ({}) });

      const result = await client._fetchWithRetry('http://example.com', {}, 3);

      expect(fetch).toHaveBeenCalledTimes(1);
      expect(result.ok).toBe(true);
    });

    test('should return error after network failures', async () => {
      fetch.mockRejectedValue(new Error('Network error'));

      await expect(
        client._fetchWithRetry('http://example.com', {}, 0)
      ).rejects.toThrow('Network error');

      expect(fetch).toHaveBeenCalledTimes(1);
    });
  });
});
